app/src/main/java/ch/epfl/sdp/connectivity/ConcreteConnectivityBroker.java

```diff
@@ -0,0 +1,135 @@
+
+    private boolean hasInternetConnection = false;
+
+    private LocationManager locationManager;
+    private Context context;
+
+    private ConnectivityManager.NetworkCallback internetNetworkCallback = new ConnectivityManager.NetworkCallback() {
+        @Override
+        public void onCapabilitiesChanged(@NonNull Network network, @NonNull NetworkCapabilities networkCapabilities) {
+            super.onCapabilitiesChanged(network, networkCapabilities);
+            if (networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)) {
+                hasInternetConnection = true;
+
+                notifyObservers(INTERNET);
+            }
+        }
+
+        @Override
+        public void onLost(@NonNull Network network) {
+            super.onLost(network);
+            hasInternetConnection = false;
+
+            notifyObservers(INTERNET);
+        }
+    };
+
+    public ConcreteConnectivityBroker(LocationManager locationManager, Context context) {
+        this.locationManager = locationManager;
+        this.context = context;
+
+        ConnectivityManager cm = (ConnectivityManager) context.getSystemService(CONNECTIVITY_SERVICE);
+        if (cm != null) {
+            cm.registerDefaultNetworkCallback(internetNetworkCallback);
+        }
+    }
+
+    @Override
+    public boolean isProviderEnabled(@NonNull Provider provider) {
+        assert provider != null;
+        switch (provider) {
+            case GPS:
+                return locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
+            case INTERNET:
+                return hasInternetConnection;
+            default:
+                throw new IllegalArgumentException("Invalid provider");
+        }
+    }
+
+    @SuppressLint("MissingPermission")
+    @Override
+    public boolean requestLocationUpdates(@NonNull Provider provider, long minTimeDelay, float minSpaceDist, LocationListener listener) {
+        assert provider != null;
+        if (provider != GPS) {
+            throw new IllegalArgumentException(String.format("Provider %s is not yet supported", provider));
+        }
+
+        if (hasPermissions(GPS)) {
+            locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, minTimeDelay, minSpaceDist, listener);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public void removeUpdates(LocationListener listener) {
+        locationManager.removeUpdates(listener);
+    }
+
+    @SuppressLint("MissingPermission")
+    @Override
+    public Location getLastKnownLocation(@NonNull Provider provider) {
+        assert provider != null;
+        if (provider == GPS && hasPermissions(provider)) {
+            return locationManager.getLastKnownLocation(LocationManager.GPS_PROVIDER);
+        }
+        return null;
+    }
+
+    @Override
+    public boolean hasPermissions(Provider provider) {
+        if (provider != GPS) {
```

Comment:
This if block can be removed since the exception is already thrown in the switch default case

Reply:
True